### PR TITLE
Add an executable for localizing multiple images

### PR DESCRIPTION
**Pull request by sebastienchappuis**
As mentionned in issue #618, this is an executable which localizes multiple views in an already existing reconstruction. The executable takes a directory of images as input along with the sfm_data file describing the original reconstruction, and outputs a new sfm_data file that contains all the new views with their poses and intrinsics.

Thank you in advance for your feedback !

### CODE REVIEW

File: src/software/Localization/main_SfM_Localization_MultipleImages.cpp [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    // Add the view to the sfm_container
+    views[v.id_view] = std::make_shared<View>(v);
+  }
+
+  GroupSharedIntrinsics(sfm_data);
+  
+  std::cout << " Total poses found : " << vec_found_poses.size() << "/" << vec_image_new.size() << endl;
+
+  // Export the found camera position in a ply.
+  const std::string out_file_name = stlplus::create_filespec(sOutDir, "found_pose_centers", "ply");
```
Comment (pmoulon) [inline src/software/Localization/main_SfM_Localization_MultipleImages.cpp]: You can use the existing function to do that (save all the camera positions):

FYI:
(**Save** can export scene in PLY and you can choose which information you want to export):
```
Save(
     sfm_data,
      out_file_name .c_str(),
     ESfM_Data(EXTRINSICS)));
```
If you want to save only the new found poses you can use this function:
https://github.com/openMVG/openMVG/blob/master/src/software/SfM/SfMPlyHelper.hpp#L21
```
plyHelper::exportToPly(vec_found_poses, out_file_name);
```

File: src/software/Localization/main_SfM_Localization_MultipleImages.cpp [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  // export also as ply
+  if (!Save(
+    sfm_data,
+    stlplus::create_filespec( sOutDir, "sfm_data_expanded.ply" ).c_str(),
+    ESfM_Data(ALL)))
+  {
+    return EXIT_FAILURE;
+  }
+
+  return EXIT_FAILURE;
```
Comment (pmoulon) [inline src/software/Localization/main_SfM_Localization_MultipleImages.cpp]: I think the binary must return EXIT_SUCCESS if everything ran fine

File: src/software/Localization/main_SfM_Localization_MultipleImages.cpp [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        KRt_From_P(matching_data.projection_matrix, &K, &R, &t);
+
+        const double focal = (K(0,0) + K(1,1))/2.0;
+        const Vec2 principal_point(K(0,2), K(1,2));
+        optional_intrinsic = std::make_shared<cameras::Pinhole_Intrinsic_Radial_K3>(
+          imageGray.Width(), imageGray.Height(),
+          focal, principal_point(0), principal_point(1));
+
+      }
+      sfm::SfM_Localizer::RefinePose
```
Comment (pmoulon) [inline src/software/Localization/main_SfM_Localization_MultipleImages.cpp]: we must check the return boolean value, BA can failed to optimize the poses for an unknown reason...

**Comment by sebastienchappuis** [comment]
I did a few corrections. I will soon fuse the two binaries (main_SfM_Localization and main_SfM_Localization_MultipleImages) together.

BTW, Is it still relevant to keep the "demo" part of the main_SfM_Localization executable (from lines 227 to 306)?

**Comment by pmoulon** [comment]
I saw that the CR (Code Review) points are all covered.
You are right, you can merge the two binary (main_SfM_Localization_MultipleImages => main_SfM_Localization)
- [ ] It will require to update the [rst doc](https://github.com/openMVG/openMVG/blob/develop/docs/sphinx/rst/software/localization/localization.rst)
- [ ] the demo mode can be removed
   - This demo mode was an easy way for me to check that the code was working as expected thanks to this mode, since the new camera position must be as close as possible as the original ones.
   - Note: this demo mode can be reproduced within your new binary by removing the set_difference to detect the existing views.

**Comment by sebastienchappuis** [comment]
All right, I fused the two localization executables into one, got rid of the "demo" mode, updated the documentation and fixed a few things !

**Comment by pmoulon** [comment]
Thank you @sebastienchappuis 
I will test it on some scenes and then perform the merge :v:

File: src/software/Localization/main_SfM_Localization.cpp [hunk 4]
```diff
@@ -22,6 +22,17 @@ using namespace openMVG::sfm;
 
 #include <cstdlib>
 
+// Naive function for finding the biggest common root dir from two paths
+std::string FindCommonRootDir(const std::string & dir1, const std::string & dir2)
```
Comment (sebastienchappuis) [inline src/software/Localization/main_SfM_Localization.cpp]: Just a note : I expected to find a function in stlplus that would do this but apparently it does not exist ? I might have missed it though !

Comment (pmoulon) [inline src/software/Localization/main_SfM_Localization.cpp]: You're right, there is not such function.
There is only the function to compare pathes between platforms stlplus::path_compare

**Comment by pmoulon** [comment]
PR merged.
Thank you @sebastienchappuis, hoping this is the start of a long series of future PR ;-)

**Comment by sebastienchappuis** [comment]
> PR merged.
Thank you @sebastienchappuis, hoping this is the start of a long series of future PR ;-)

could very well be :) 
thank you for your help !

**Pull request merged**